drivers:dac: added ltc2672 support.

jsanbuen opened the pull request:
Added driver implementation for ltc2672.

More information: https://www.analog.com/media/en/technical-documentation/data-sheets/ltc2672.pdf

Signed-off-by: Jose Ramon San Buenaventura <jose.sanbuenaventura@analog.com>

amiclaus commented on drivers/dac/ltc2672/ltc2672.c:
```diff
@@ -0,0 +1,515 @@
+		out_range_sum += device->out_spans[i];
+	}
+
+	out_range_sum /= 5;
```
a macro would be nice for `5`

amiclaus commented on drivers/dac/ltc2672/ltc2672.c:
```diff
@@ -0,0 +1,515 @@
+	if(ret)
+		return ret;
+
+	if(ch_span == LTC2672_VMINUS_VREF) {
```
`switch... case`?

amiclaus commented on drivers/dac/ltc2672/ltc2672.c:
```diff
@@ -0,0 +1,515 @@
+	if(ch_span < LTC2672_OFF || ch_span > LTC2672_4800VREF)
+		return -EINVAL;
+
+	for(int i = 0; i < 5; i++)
```
nit: `for (`

amiclaus commented on drivers/dac/ltc2672/ltc2672.c:
```diff
@@ -0,0 +1,515 @@
+	uint16_t span = 0;
+	uint32_t command;
+
+	if(ch_span < LTC2672_OFF || ch_span > LTC2672_4800VREF)
```
nit: `if (`

amiclaus commented on drivers/dac/ltc2672/ltc2672.c:
```diff
@@ -0,0 +1,515 @@
+					     (uint32_t)current_code);
+
+	ret = ltc2672_transaction(device, command, false);
+	if(ret)
```
`if (` check throughout the entire code.

amiclaus commented on drivers/dac/ltc2672/ltc2672.h:
```diff
@@ -0,0 +1,257 @@
+
+/* Device Family */
+enum ltc2672_device_id {
+	LTC2672_12 = 0,
```
no need for explicit initialization

amiclaus commented on drivers/dac/ltc2672/ltc2672.h:
```diff
@@ -0,0 +1,257 @@
+	LTC2672_MUX_IOUT3,
+	LTC2672_MUX_IOUT4,
+	LTC2672_MUC_VCC,
+	LTC2672_MUX_VREF = 0x08,
```
in this case you either initialize all members of the enum, or none of them.

buha replied:
It's weird to specify a value for all of them. Just specify the value for those that create a gap like LTC2672_MUX_VREF which is +2 compared to  previous one LTC2672_MUC_VCC and leave the others without explicit values.

buha replied:
So I think @jsanbuen was right.

jsanbuen commented:
Changelog/s:
- added space in if and for as suggested by reviewer.
- completed initialization of enum members and removal of some explicit enum initialization.
- added LTC2672_TOTAL_CHANNELS as macro for 5.
- changed some parts to switch case statements from if-else if-else statements.

jsanbuen commented:
Changelog/s:
-used NO_OS_BIT function for applicable instances.

jsanbuen commented:
Changelog/s:
-corrected astyle formatting from affected lines.
-corrected paranthesis mismatch from changing to NO_OS_BIT function.
-removed duplicate setup_toggle_channel function in header file.
-corrected ltc2672_transaction function prototype by using only struct ltc2672_dev * instead of including variable name.

amiclaus commented on drivers/dac/ltc2672/ltc2672.c:
```diff
@@ -0,0 +1,532 @@
+		for (int i = 0; i < LTC2672_TOTAL_CHANNELS; i++)
+			device->max_currents[i] = device->neg_current;
+		break;
+
```
nit: no need for these extra lines

amiclaus commented on drivers/dac/ltc2672/ltc2672.c:
```diff
@@ -0,0 +1,532 @@
+{
+	uint32_t command;
+
+	if (mask > 15)
```
what's the meaning of `15`? Add a suggestive macro since all the defined masks in the header file seem to be greater than 15.

jsanbuen commented:
Changelog/s:
-added macros for 15 (LTC2672_MAX_CONFIG_MASK) and 31 (LTC2672_MAX_TOGGLE_MASK).
-removed extra spaces in switch statements in lines 304 - 322 and 256 - 271.

buha commented on drivers/dac/ltc2672/ltc2672.c:
```diff
@@ -0,0 +1,526 @@
+	descriptor->active_device = init_param->active_device;
+	descriptor->prev_command = 0x00000000;
+	descriptor->global_toggle = false; // default toggle is false
+	descriptor->neg_current = init_param->neg_current;
```
You seem to set a device property from init_param but where is this actually set in the device register ?

buha commented on drivers/dac/ltc2672/ltc2672.c:
```diff
@@ -0,0 +1,526 @@
+	else
+		bytes = 3;
+
+	for (int i = bytes; i > 0; i--)
```
you have `bytes` declared above as uint16_t, why declare another one ?
also, this inline declaration is tricky, i don't think it works with all C compilers so best to drop it

jsanbuen commented:
Changelog/s:
- added new macro definitions for items such as off current, 300mA current, fixed -80mA current span etc.
- removed some parameters (negative current, since fixed to -80mA upon test) and other zero initialized values due to calloc in init function.
- revised all inline declarations in for loops.
- enum elements were returned to original state where only elements that create a gap are explicitly declared.

jsanbuen commented:
Changelog/s:
-Astyle formatting correction on ltc2672.c file

jsanbuen commented:
Changelog/s:
- removed extra spaces in switch statements

buha commented on drivers/dac/ltc2672/ltc2672.h:
```diff
@@ -0,0 +1,251 @@
+	/* SPI descriptor */
+	struct no_os_spi_desc *comm_desc;
+	/* Device Variant indicator */
+	enum ltc2672_device_id active_device;
```
`active_device` is kinda long...
I know this is semantics, but `id` would make more sense.

jsanbuen replied:
Changed active_device property to id as suggested on recent push.

buha reviewed (commented):
Looks good now, that minor change and we can merge.
